File: Evaluator.cpp
Sol eval(Atom *a, Sol self, Sol contextClass, Environment environment) {
 tailRecursionStart:
#ifdef Log
  evaluationStackIndent();
  LogEvaluation<<a->toString()<<LogEnd;
#endif
  if(a == nullptr) {
    return nullptr;
  }
  Sol returnValue;
  switch(a->type) {
  case Atom::Type::message: {
    GlobalEvaluationStackCount++;
    Sol receiver, klass;
    vector<Sol> arguments;
    if(a->message.receiver->type == Atom::Type::identifier and
       a->message.receiver->identifier == "super") {
      if(contextClass == nullptr)
	throw RuntimeException(/*a, */"Cannot refer to super while not in a valid context.");
      receiver = self;
      for(Atom *argument : a->message.arguments)
	arguments.push_back(eval(argument, self, contextClass, environment));
      klass = contextClass->getVariable("superclass");
    } else {
      receiver = eval(a->message.receiver, self, contextClass, environment);
      for(Atom *argument : a->message.arguments)
	arguments.push_back(eval(argument, self, contextClass, environment));
      klass = receiver->getClass();
    }
    try {
      GlobalEvaluationStackCount--;
      return receiver->sendWER(a->message.message, arguments, klass);	
    } catch(EvaluationRequest request) {
      a            = std::move(request.atom);
      self         = std::move(request.self);
      contextClass = std::move(request.contextClass);
      environment  = std::move(request.environment);
      
      LogEvaluation<<"Catching "<<LogEnd;
      goto tailRecursionStart;
    }
  }
  case Atom::Type::identifier:
    if(a->identifier == "self")
      if(self == nullptr)
	throw RuntimeException(/*a, */"Cannot refer to self while not in a valid context.");
      else
	return self;
    else if(a->identifier == "super")
      throw RuntimeException(/*a, */"Cannot refer to super as a standalone variable.");
    else if(a->identifier == "nothing")
      return (Sol)nullptr;
    else if(environment->has(a->identifier))
      return environment->get(a->identifier);
    else if(self != nullptr and self->hasVariable(a->identifier))
      return self->getVariable(a->identifier);
    else
      throw EnvironmentException(environment, a->location, a->identifier);
  case Atom::Type::assignment: {
    GlobalEvaluationStackCount++;
    Sol value = eval(a->assignment.value, self, contextClass, environment);
    GlobalEvaluationStackCount--;
    string variable = a->assignment.variable;
    if(self != nullptr and self->hasVariable(variable))
      self->setVariable(variable, value);
    else if(environment->has(variable))
      environment->set(variable, value);
    else
      environment->define(variable, value);
    return self;
  }
  case Atom::Type::sequence: {
    int i;
    GlobalEvaluationStackCount++;
    for(i=0; i<a->sequence.size()-1; i++)
      eval(a->sequence[i], self, contextClass, environment);
    GlobalEvaluationStackCount--;
    a = a->sequence[i];
    goto tailRecursionStart;
  }
  case Atom::Type::lambda: {
    return SolCreate(Lambda, new LambdaStructure(a->lambda.body,
						 a->lambda.parameters,
						 self,
						 environment));
  }
  case Atom::Type::vector: {
    vector<Sol> elements;
    GlobalEvaluationStackCount++;
    for(Atom *element : a->vectorElements) {
      Sol elementObject = eval(element, self, contextClass, environment);
      elements.push_back(elementObject);
    }
    GlobalEvaluationStackCount--;
    return SolCreate(Vector, new vector<Sol>(elements));
  }
  case Atom::Type::symbol:
    return SolCreate(Symbol, new string(a->symbol)); 
  case Atom::Type::integer:
    return SolCreate(Integer, new int(a->integerValue)); 
  case Atom::Type::double:
    return SolCreate(Double, new double(a->doubleValue)); 
  case Atom::Type::character:
    return SolCreate(Character, new char(a->characterValue)); 
  case Atom::Type::string:
    return SolCreate(String, new string(a->stringValue)); 
  default:
    cerr<<"Fatal error, unrecognized Atom."<<endl;
  }
}
inline Sol eval(Atom *atom, Environment environment) {
  return eval(atom, nullptr, nullptr, environment);
}

File: Evaluator.java
import java.util.*;

public class Evaluator {
    public static Sol eval(Atom atom) throws Exception {
	return eval(atom, null, null, null);
    }
    public static Sol eval(Atom atom, Environment environment) throws Exception {
	return eval(atom, null, null, environment);
    }
    public static Sol eval(Atom atom, Sol self, Sol contextClass, Environment environment) throws Exception {
	while(true) 
	    if(atom == null)
		return null;
	    else if(atom instanceof MessageAtom) {
		MessageAtom messageAtom = (MessageAtom)atom;
		Atom messageReceiver = messageAtom.getReceiver();
		Sol receiver, clazz;
		if(messageReceiver instanceof IdentifierAtom &&
		   ((IdentifierAtom)messageReceiver).getIdentifier().equals("super")) {
		    if(contextClass == null)
			throw new RuntimeException("Cannot refer to super while not in a valid context.");
		    receiver = self;	    
		    clazz = contextClass.getVariable("superclass");
		} else {
		    receiver = eval(messageReceiver, self, contextClass, environment);
		    clazz = receiver == null ? Classes.Nothing.get() : receiver.getClazz();
		}

		Atom[] messageArguments = messageAtom.getArguments();
		Sol[] arguments = new Sol[messageArguments.length];
		for(int i=0; i<messageArguments.length; i++)
		    arguments[i] = eval(messageArguments[i], self, contextClass, environment);

		Message message = messageAtom.getMessage();
		try {
		    return Sol.sendWER(receiver, message, arguments, clazz);
		} catch(EvaluationRequest request) {
		    atom = request.atom;
		    self = request.self;
		    contextClass = request.contextClass;
		    environment = request.environment;
		    continue;
		}
	    } else if(atom instanceof IdentifierAtom) {
		IdentifierAtom identifierAtom = (IdentifierAtom)atom;
		String identifier = identifierAtom.getIdentifier();
		if(identifier.equals("self"))
		    if(self == null)
			throw new RuntimeException(atom, "Cannot refer to self while not in a valid context.");
		    else
			return self;
		else if(identifier.equals("super"))
		    throw new RuntimeException("Cannot refer to super as a standalone variable.");
		else if(identifier.equals("nothing"))
		    return null;
		else if(environment.has(identifier))
		    return environment.get(identifier);
		else if(self != null && self.hasVariable(identifier))
		    return self.getVariable(identifier);
		else
		    throw new RuntimeException(atom, "Undefined variable \"" + identifier + "\".");
	    } else if(atom instanceof SequenceAtom) {
		SequenceAtom sequenceAtom = (SequenceAtom)atom;
		Atom[] sequence = sequenceAtom.getSequence();
		for(int i=0; i<sequence.length-1; i++)
		    eval(sequence[i], self, contextClass, environment);

		atom = sequence[sequence.length-1];
		continue;
	    } else if(atom instanceof AssignmentAtom) {
		AssignmentAtom assignmentAtom = (AssignmentAtom)atom;
		String variable = assignmentAtom.getVariable();
		Atom expression = assignmentAtom.getExpression();
		Sol value = eval(expression, self, contextClass, environment);
		if(self != null && self.hasVariable(variable))
		    self.setVariable(variable, value);
		else if(environment.has(variable))
		    environment.set(variable, value);
		else
		    environment.define(variable, value);
		return self;
	    } else if(atom instanceof LambdaAtom) {
		LambdaAtom lambda = (LambdaAtom)atom;
		Atom body = lambda.getBody();
		String[] parameters = lambda.getParameters();
		return new Sol(Classes.Lambda, new Lambda(body, parameters,
							  self, environment));
	    } else if(atom instanceof VectorAtom) {
		VectorAtom vectorAtom = (VectorAtom)atom;
		Atom[] vector = vectorAtom.getVector();
		Sol[] elements = new Sol[vector.length];
		for(int i=0; i<vector.length; i++)
		    elements[i] = eval(vector[i], self, contextClass, environment);
		return new Sol(Classes.Vector, new ArrayList<Sol>(Arrays.asList(elements)));
	    } else if(atom instanceof IntegerAtom) {
		return new Sol(Classes.Integer, ((IntegerAtom)atom).getInteger());
	    } else if(atom instanceof DoubleAtom) {
		return new Sol(Classes.Double, ((DoubleAtom)atom).getDouble());
	    } else if(atom instanceof CharacterAtom) {
		return new Sol(Classes.Character, ((CharacterAtom)atom).getCharacter());
	    } else if(atom instanceof StringAtom) {
		return new Sol(Classes.String, ((StringAtom)atom).getString());
	    } else if(atom instanceof MessageSymbolAtom) {
	        return new Sol(Classes.Message, ((MessageSymbolAtom)atom).getMessage());
	    } else {
		System.out.println("Fatal error, unrecognized Atom.");
	    }
    }
}
